// @mapquest/io/dist/src/speed-alerts/service.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var service_1 = require('../generic/service');
var environment_1 = require('../environment');
var _1 = require('../security/');
var SpeedAlertService = (function (_super) {
    __extends(SpeedAlertService, _super);
    function SpeedAlertService() {
        _super.apply(this, arguments);
    }
    SpeedAlertService.prototype.getAPIHost = function () {
        return environment_1.CURRENT_ENVIRONMENT['et-rules'];
    };
    SpeedAlertService.prototype.getRootPath = function () {
        return "/speed-alerts/v1/" + this.applicationContext.applicationId;
    };
    SpeedAlertService.prototype.create = function (model, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = this.getAPIHost() + this.getRootPath();
        return this.http.post(resourceURL, model, options);
    };
    SpeedAlertService.prototype.get = function (id, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath() + "/" + id;
        return this.http.get(resourceURL, options);
    };
    SpeedAlertService.prototype.list = function (criteria, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath();
        if (criteria && criteria.clientId)
            resourceURL += '?clientId=' + criteria.clientId;
        return this.http.get(resourceURL, options);
    };
    SpeedAlertService.prototype.delete = function (id, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath() + "/" + id;
        return this.http.delete(resourceURL, options);
    };
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object, Object]), 
        __metadata('design:returntype', Promise)
    ], SpeedAlertService.prototype, "create", null);
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [String, Object]), 
        __metadata('design:returntype', Promise)
    ], SpeedAlertService.prototype, "get", null);
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object, Object]), 
        __metadata('design:returntype', Promise)
    ], SpeedAlertService.prototype, "list", null);
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [String, Object]), 
        __metadata('design:returntype', Promise)
    ], SpeedAlertService.prototype, "delete", null);
    return SpeedAlertService;
}(service_1.GenericService));
exports.SpeedAlertService = SpeedAlertService;
