// @mapquest/io/src/index.ts
import { ApplicationContext } from './application-context';
import { ApplicationAccess } from './security/';
import { GenericService } from './generic/service';
import { IRule } from './rules/';
import { IGeofence, GeofenceService } from './geofences/';
import { IObservation, ObservationPayload } from './observations/';
import { ISpeedAlert, SpeedAlertService } from './speed-alerts/';
import { ITransition, GeofenceTransition, EnterTransition, ExitTransition, DwellTransition, SpeedTransition } from './transitions/';
import { IApplication } from './applications/';
import { IClient, ClientService } from './clients/';
import { IRegion, CircularRegion } from './regions/';
import { CURRENT_ENVIRONMENT } from './environment';
import { Webhook, SlackNotification, IReaction } from './reactions/';

let Configuration = {
    secret: process.env.MQ_IO_APPLICATION_SECRET
};

export function getConfiguration() {
    return Configuration;
}

export {
    ApplicationContext,
    ApplicationAccess,
    GenericService,
    IRule,
    IGeofence, GeofenceService,
    IObservation, ObservationPayload,
    ISpeedAlert, SpeedAlertService,
    IRegion, CircularRegion,
    ITransition, GeofenceTransition, EnterTransition, ExitTransition, DwellTransition, SpeedTransition,
    IApplication,
    IClient, ClientService, 
    Webhook, SlackNotification, IReaction,

    CURRENT_ENVIRONMENT
};
export * from './http/';
