// @mapquest/io/src/geofences/service.ts
import { GenericService } from '../generic';
import { IGeofence } from './model';

import { CURRENT_ENVIRONMENT } from '../environment';
import { ApplicationAccess } from '../security/';

export class GeofenceService extends GenericService<IGeofence> {
    getAPIHost() {
        return CURRENT_ENVIRONMENT['et-rules'];
    }

    getRootPath() {
        return `/geofences/v1/${this.applicationContext.applicationId}`;
    }

    @ApplicationAccess
    create(model: IGeofence, options:any = {}): Promise<IGeofence> {
        let resourceURL = this.getAPIHost() + this.getRootPath();
        return this.http.post<IGeofence>(resourceURL, model, options);
    }

    @ApplicationAccess
    get(id: string, options:any = {}): Promise<IGeofence> {
      let resourceURL = `${this.getAPIHost()}${this.getRootPath()}/${id}`;
      return this.http.get<IGeofence>(resourceURL, options) as Promise<IGeofence>;
    }

    @ApplicationAccess
    list(criteria?: any, options:any = {}): Promise<IGeofence[]> {
        let resourceURL = this.getAPIHost() + this.getRootPath();
        if ( criteria && criteria.clientId ) resourceURL += '?clientId=' + criteria.clientId;
        return this.http.get<IGeofence[]>(resourceURL, options) as Promise<IGeofence[]>;
    }

    @ApplicationAccess
    delete(id: string, options:any = {}): Promise<boolean> {
      let resourceURL = `${this.getAPIHost()}${this.getRootPath()}/${id}`;
      return this.http.delete<boolean>(resourceURL, options);
    }
}
