// @mapquest/io/dist/src/observations/model.js
"use strict";
var StartedRoute = (function () {
    function StartedRoute(routeSessionId) {
        this.routeSessionId = routeSessionId;
        this.type = 'startedroute';
    }
    return StartedRoute;
}());
exports.StartedRoute = StartedRoute;
var AbandonedRoute = (function () {
    function AbandonedRoute(routeSessionId) {
        this.routeSessionId = routeSessionId;
        this.type = 'abandonedroute';
    }
    return AbandonedRoute;
}());
exports.AbandonedRoute = AbandonedRoute;
var CompletedRoute = (function () {
    function CompletedRoute(routeSessionId) {
        this.routeSessionId = routeSessionId;
        this.type = 'completedroute';
    }
    return CompletedRoute;
}());
exports.CompletedRoute = CompletedRoute;
var ObservationPayload = (function () {
    function ObservationPayload(clientId, observations, sessionId) {
        this.clientId = clientId;
        this.observations = observations;
        this.sessionId = sessionId;
    }
    ObservationPayload.prototype.toBody = function () {
        var body = {
            observations: this.observations
        };
        if (this.sessionId) {
            body.sessionId = this.sessionId;
        }
        return body;
    };
    return ObservationPayload;
}());
exports.ObservationPayload = ObservationPayload;
