// @mapquest/io/dist/src/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var application_context_1 = require('./application-context');
exports.ApplicationContext = application_context_1.ApplicationContext;
var _1 = require('./security/');
exports.ApplicationAccess = _1.ApplicationAccess;
var service_1 = require('./generic/service');
exports.GenericService = service_1.GenericService;
var _2 = require('./geofences/');
exports.GeofenceService = _2.GeofenceService;
var _3 = require('./observations/');
exports.ObservationPayload = _3.ObservationPayload;
var _4 = require('./speed-alerts/');
exports.SpeedAlertService = _4.SpeedAlertService;
var _5 = require('./transitions/');
exports.EnterTransition = _5.EnterTransition;
exports.ExitTransition = _5.ExitTransition;
exports.DwellTransition = _5.DwellTransition;
exports.SpeedTransition = _5.SpeedTransition;
var _6 = require('./clients/');
exports.ClientService = _6.ClientService;
var _7 = require('./regions/');
exports.CircularRegion = _7.CircularRegion;
var environment_1 = require('./environment');
exports.CURRENT_ENVIRONMENT = environment_1.CURRENT_ENVIRONMENT;
var _8 = require('./reactions/');
exports.Webhook = _8.Webhook;
exports.SlackNotification = _8.SlackNotification;
var Configuration = {
    secret: process.env.MQ_IO_APPLICATION_SECRET
};
function getConfiguration() {
    return Configuration;
}
exports.getConfiguration = getConfiguration;
__export(require('./http/'));
