// @mapquest/io/dist/src/clients/service.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var service_1 = require('../generic/service');
var observations_1 = require('../observations');
var environment_1 = require('../environment');
var _1 = require('../security/');
var ClientService = (function (_super) {
    __extends(ClientService, _super);
    function ClientService() {
        _super.apply(this, arguments);
    }
    ClientService.prototype.getAPIHost = function () {
        return environment_1.CURRENT_ENVIRONMENT['et-devices'];
    };
    ClientService.prototype.getRootPath = function () {
        return "/clients/v1/" + this.applicationContext.applicationId;
    };
    ClientService.prototype.create = function (model, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath() + "/" + model.clientId;
        return this.http.post(resourceURL, model, options);
    };
    ClientService.prototype.update = function (model, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath() + "/" + model.clientId;
        return this.http.patch(resourceURL, model, options);
    };
    ClientService.prototype.postObservations = function (payload, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath() + "/" + payload.clientId + "/observations";
        return this.http.post(resourceURL, payload.toBody(), options);
    };
    ClientService.prototype.reset = function (model, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath() + "/" + model.clientId + "/reset/secret";
        return this.http.post(resourceURL, null, options);
    };
    ClientService.prototype.list = function (criteria, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath();
        return this.http.get(resourceURL, options);
    };
    ClientService.prototype.get = function (id, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath() + "/" + id;
        return this.http.get(resourceURL, options);
    };
    ClientService.prototype.delete = function (id, options) {
        if (options === void 0) { options = {}; }
        var resourceURL = "" + this.getAPIHost() + this.getRootPath() + "/" + id;
        return this.http.delete(resourceURL, options);
    };
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object, Object]), 
        __metadata('design:returntype', Promise)
    ], ClientService.prototype, "create", null);
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object, Object]), 
        __metadata('design:returntype', Promise)
    ], ClientService.prototype, "update", null);
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [observations_1.ObservationPayload, Object]), 
        __metadata('design:returntype', Promise)
    ], ClientService.prototype, "postObservations", null);
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object, Object]), 
        __metadata('design:returntype', void 0)
    ], ClientService.prototype, "reset", null);
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object, Object]), 
        __metadata('design:returntype', Promise)
    ], ClientService.prototype, "list", null);
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [String, Object]), 
        __metadata('design:returntype', Promise)
    ], ClientService.prototype, "get", null);
    __decorate([
        _1.ApplicationAccess, 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [String, Object]), 
        __metadata('design:returntype', Promise)
    ], ClientService.prototype, "delete", null);
    return ClientService;
}(service_1.GenericService));
exports.ClientService = ClientService;
